### PR TITLE
Premium users stuck when instance stopped, but they log in again

### PR DESCRIPTION
**Pull request by milesAraya**
#### Summary
- Premium users whose dedicated instance was stopped for inactivity get stuck: the assign endpoint finds an active RDS assignment and returns it, but the instance is stopped. The user cannot use the platform until cleanup removes the stale assignment record (runs hourly), after which a fresh instance is provisioned.
- The fix adds an EC2 state check in the assign path: when the user's assigned instance is stopped, the endpoint restarts it inline, waits for EC2 running + ECS task readiness, and returns the assignment. For terminated or missing instances, the stale assignment is removed and the user falls through to fresh assignment.
- The existing migration and autoscaling-pool paths are unchanged.

#### Design Decisions
- **EC2 state check before returning existing assignment** -- The DB `instance_state` field can be stale (updated only on explicit transitions, not when the monitoring Lambda stops idle instances). Querying EC2 `describe_instances` is the only reliable way to know the real state. This adds one API call per assign request when the user already has an assignment, which is acceptable. Reuses the `ec2` client created at the top of `assign_premium_user` rather than creating a second client.
- **Inline restart with 2-minute EC2 wait + 2-minute ECS readiness** -- Stopped instances resume in ~30-60s. The EC2 waiter uses `Delay=5, MaxAttempts=24` (~2 min max) matching the existing `start_standby_instance()` pattern. ECS readiness uses `max_wait_seconds=120` to allow the ECS agent to re-register and start the premium task. This is faster than a full cold-start assignment (Priority 3-5) and preserves the user's existing ALB rule and target group.
- **STOPPING → STOPPED wait before restart** -- An instance in `stopping` state cannot be started directly (AWS throws `IncorrectInstanceState`). The code waits for the instance to reach `stopped` via `instance_stopped` waiter before issuing `start_instances`.
- **Stale assignment cleanup for terminated/missing instances** -- If the instance is terminated, shutting down, or not found in EC2, the assignment is removed and `existing_assignment` is set to `None` so the code falls through to the fresh assignment flow (Priority 1-5). `existing_assignment` is set to `None` **before** calling `remove_user_assignment()` so that if the removal throws, the stale record is never returned to the user.
- **Guard on `existing_assignment` for migration and return paths** -- After cleanup, `existing_assignment` is set to `None` to prevent the downstream code from returning a deleted assignment or triggering migration for a non-existent record. The catch-all `except Exception` also sets `existing_assignment = None` to avoid returning a potentially broken assignment.
- **Non-NotFound `ClientError` re-raised** -- Unknown EC2 errors (throttling, permissions) are re-raised to the outer handler rather than silently swallowed, since they may indicate a systemic issue.
- **DB update via `@with_transaction`** -- The instance state update after restart uses a new `_update_instance_state_to_running()` function decorated with `@with_transaction`, consistent with all other DB writes in the file (e.g., `_remove_user_assignment_transaction`, `_store_user_assignment_transaction`).

#### Scenario (reported by coworker)
1. Premium user is assigned to a dedicated instance, assignment stored in RDS
2. Instance is stopped by the idle-timeout monitoring Lambda
3. User returns and calls assign — endpoint finds active RDS record, returns it pointing to stopped instance
4. User is stuck until hourly cleanup deletes the stale assignment
5. Only then can a fresh instance be provisioned

#### Expected behavior after fix
- Step 3: assign detects stopped instance via EC2 API, restarts it, waits for readiness, returns assignment with `assignment_source: "restarted_instance"`
- User sees ~1-2 minute wait instead of being stuck indefinitely

### References
- Related: commit 6247a9680 (autoscaling-pool inline migration fix)
- PULL_REQUEST.md (autoscaling-pool loop fix, same branch)

#### Files changed
- `infrastructure/terraform/premium_manager_package/premium_manager.py` -- Add EC2 state check in assign path: restart stopped instances (with STOPPING wait), clean up terminated/missing instances, guard downstream paths with `existing_assignment` nullability. Extract `_update_instance_state_to_running()` as `@with_transaction` function. Re-raise non-NotFound `ClientError`s. Reuse existing `ec2` client instead of creating a second one.
- `studio/tests/infrastructure/test_premium_manager.py` -- Add 4 new tests for stopped/stopping/terminated/not-found instance handling; fix `test_assign_event` mock sequence off-by-one from `restore_pending_release` fetchone added in 6247a9680

### Manual Testcases
1. **Premium user with stopped instance calls assign**
   - Precondition: User has active assignment in RDS, but the EC2 instance is stopped (e.g., stopped by idle monitoring)
   - Action: User logs in or refreshes, triggering assign endpoint
   - Expected: Instance restarts, ECS task becomes ready, response returns `assignment_source: "restarted_instance"` with the same instance_id
2. **Premium user with terminated instance calls assign**
   - Precondition: User has active assignment in RDS, but the EC2 instance was terminated
   - Action: User logs in, triggering assign endpoint
   - Expected: Stale assignment removed, user falls through to fresh assignment flow (Priority 1-5)
3. **Premium user with running instance calls assign (no change)**
   - Precondition: User has active assignment, instance is running
   - Action: User calls assign
   - Expected: Existing assignment returned as before with `assignment_source: "existing"`
4. **Premium user with stopping instance calls assign**
   - Precondition: User has active assignment in RDS, EC2 instance is in `stopping` state (e.g., idle-timeout Lambda just triggered)
   - Action: User logs in, triggering assign endpoint
   - Expected: Endpoint waits for instance to reach `stopped`, then restarts it, waits for ECS readiness, returns `assignment_source: "restarted_instance"`
5. **Autoscaling-pool user calls assign (no change)**
   - Precondition: User assigned to autoscaling-pool
   - Action: User calls assign
   - Expected: Inline migration or async migration triggers as before (EC2 state check is skipped for autoscaling-pool)
6. `make test_lambda` -- all tests pass

### Unit, Integration, Contract Test Coverage
- `make test_lambda` -- all passed, 0 failed
- New tests:
  - `test_stopped_instance_restarted_on_assign` -- Verifies EC2 restart flow, ECS readiness wait, DB state update via `_update_instance_state_to_running`, and `assignment_source: "restarted_instance"` response
  - `test_stopping_instance_waits_then_restarts` -- Verifies STOPPING → wait for stopped → restart flow, asserts `instance_stopped` waiter is called before `start_instances`
  - `test_terminated_instance_triggers_fresh_assignment` -- Verifies stale assignment cleanup and fall-through to fresh assignment
  - `test_ec2_not_found_cleans_up_stale_assignment` -- Verifies `InvalidInstanceID.NotFound` error handling and stale assignment cleanup

#### Difficulties (if any)
- None

#### Risk Assessment
| Area | Risk | Notes |
|------|------|-------|
| Assign endpoint latency | Medium | Restarting a stopped instance adds ~1-2 min to the assign response. A `stopping` instance adds additional wait time. This is expected and preferable to the user being stuck. The 120s ECS readiness timeout is a worst case; typical restart is 30-60s. |
| Lambda timeout | Medium | Worst case: STOPPING wait (~2 min) + start wait (~2 min) + ECS readiness (~2 min) = ~6 min. Confirm Lambda timeout is >= 6 min or adjust waiter configs. |
| EC2 API call on every existing-assignment path | Low | One `describe_instances` call per assign when user already has an assignment. Well within API rate limits. |
| Stale assignment cleanup | Low | `remove_user_assignment` is an existing, tested function. `existing_assignment` is set to `None` before the call so failures in removal cannot result in returning a stale record. |
| Non-NotFound ClientError re-raised | Low | Unknown EC2 errors now propagate to the outer handler (503 response) instead of being silently logged. This surfaces systemic issues (throttling, IAM) rather than hiding them. |

### CODE REVIEW

File: infrastructure/terraform/premium_manager_package/premium_manager.py [hunk 1]
```diff
@@ -2345,8 +2359,115 @@ def assign_premium_user(
...
+                        )
+                        ec2_state = InstanceState.STOPPED
+
+                    if ec2_state == InstanceState.STOPPED:
+                        print(
+                            f"Assigned instance {existing_instance_id} is "
+                            f"stopped — restarting for user {user_id}"
+                        )
+                        ec2.start_instances(InstanceIds=[existing_instance_id])
+                        waiter = ec2.get_waiter("instance_running")
```
Comment (tsuchiyama-araya) [inline infrastructure/terraform/premium_manager_package/premium_manager.py:2610]: Maybe we can use const or enum for this?

File: infrastructure/terraform/premium_manager_package/premium_manager.py [hunk 2]
```diff
@@ -568,6 +568,20 @@ def remove_user_assignment(user_id: int):
     return _remove_user_assignment_transaction(user_id)
 
 
+@with_transaction
+def _update_instance_state_to_running(connection, instance_id: str):
+    """Update instance state to running after restart"""
+    with connection.cursor() as cursor:
+        cursor.execute(
+            """UPDATE premium_user_assignments
+               SET instance_state = 'running',
```
Comment (itutu-tienday) [inline infrastructure/terraform/premium_manager_package/premium_manager.py]: Wouldn't it be possible to use `class InstanceState` to specify this value?

File: infrastructure/terraform/premium_manager_package/premium_manager.py (hunk 1 above)
Comment (milesAraya) [reply infrastructure/terraform/premium_manager_package/premium_manager.py:2610]: This is a boto3 API string and doesn't warrant a custom constant

File: infrastructure/terraform/premium_manager_package/premium_manager.py (hunk 2 above)
Comment (milesAraya) [reply infrastructure/terraform/premium_manager_package/premium_manager.py]: Fixed: instance_state = 'running' → Use InstanceState.RUNNING

**Review by tsuchiyama-araya — APPROVED** [review]
